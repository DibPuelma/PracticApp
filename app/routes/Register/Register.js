'use strict';

import React, { Component } from 'react'
import {
  StyleSheet,
  View,
  Text,
  TextInput,
  Navigator,
  TouchableHighlight,
  TouchableOpacity,
  ActivityIndicator,
  ScrollView,
  AsyncStorage,
  Picker
} from 'react-native';

import DatePicker from 'react-native-datepicker';
import styles from './styles';
import backButtonHandler from '../../lib/backButtonHandler';
import settings from '../../config/settings'

var LoginStatus = {
  SLEEPING: 'slepping',
  REQUESTING: 'requesting',
  LOGGED: 'logged'
}

export default class Register extends Component {
  constructor(props) {
    super(props);
    this.state = {name: '', lastName: '', email: '', password: '', confirmPassword: '', gender: 'm', status: LoginStatus.SLEEPING};
    this._backToMainMenu = this._backToMainMenu.bind(this);
    this.singletonBackButtonHandler = backButtonHandler.getInstance();
  }

  componentWillMount() {
    this.singletonBackButtonHandler.addBackEvent(this._backToMainMenu);
  }

  componentWillUnmount() {
    this.singletonBackButtonHandler.removeBackEvent(this._backToMainMenu);
  }

  render() {
    return (
      <ScrollView style={styles.scrollview}>
      <View style={styles.container}>
        <View style={styles.loginColumn}>
          <View style={styles.logo} />
          <Text style={styles.title}>Registro</Text>
            { this.state.error &&
              <Text style={{color: 'red'}}>{this.state.error}</Text>
            }

            <View style={{flexDirection: 'row', marginBottom: 20}}>
              <TextInput
                maxLength = {30}
                style={{height: 40, flex:1}}
                placeholder="Nombre"
                onChangeText={(text) => this.setState({name: text})}
              />
            </View>

            <View style={{flexDirection: 'row', marginBottom: 20}}>
              <TextInput
                maxLength = {30}
                style={{height: 40, flex:1}}
                placeholder="Apellido"
                onChangeText={(text) => this.setState({lastName: text})}
              />
            </View>

            <View style={{flexDirection: 'row', marginBottom: 20}}>
              <TextInput
                maxLength = {30}
                style={{height: 40, flex:1}}
                placeholder="Correo"
                onChangeText={(text) => this.setState({email: text})}
              />
            </View>

            <View style={{flexDirection: 'row', marginLeft: 6}}>
              <Text style={{fontSize: 12, color: '#999', flex: 1}}>Género:</Text>
            </View>

            <View style={{flexDirection: 'row', marginBottom: 30}}>
              <Picker selectedValue={this.state.gender} style={{flex: 1}} onValueChange={(g) => this.setState({gender: g})}>
                <Picker.Item label="Masculino" value="m" />
                <Picker.Item label="Femenino" value="f" />
                <Picker.Item label="Otro" value="o" />
              </Picker>
            </View>

            <View style={{flexDirection: 'row', marginBottom: 20}}>
              <DatePicker
                style={{flex: 1}}
                date={this.state.birthdate}
                mode="date"
                placeholder="Fecha de nacimiento"
                format="DD/MM/YYYY"
                minDate="01/01/1900"
                maxDate="01/01/2010"
                confirmBtnText="Confirmar"
                cancelBtnText="Cancelar"
                customStyles={{
                  dateIcon: {
                    position: 'absolute',
                    left: 0,
                    top: 4,
                    marginLeft: 0
                  },
                  dateInput: {
                    marginLeft: 36
                  }
                  // ... You can check the source to find the other keys.
                }}
                onDateChange={(birthdate) => {this.setState({birthdate: birthdate})}}
              />
            </View>


            <View style={{flexDirection: 'row', marginBottom: 20}}>
              <TextInput
                maxLength = {30}
                secureTextEntry={true}
                style={{height: 40, flex:1}}
                placeholder="Contraseña"
                onChangeText={(text) => this.setState({password: text})}
              />
            </View>

            <View style={{flexDirection: 'row', marginBottom: 20}}>
              <TextInput
                maxLength = {30}
                secureTextEntry={true}
                style={{height: 40, flex:1}}
                placeholder="Confirmar Contraseña"
                onChangeText={(text) => this.setState({confirmPassword: text})}
              />
            </View>


          <View style={styles.loginButtons}>
            { this.state.status === LoginStatus.SLEEPING &&
            <TouchableHighlight style={styles.fullWidthButton} underlayColor="#66CCF5" onPress={this._login.bind(this)}>
              <Text style={styles.fullWidthButtonText}>ENTRAR</Text>
            </TouchableHighlight>
            }
            { this.state.status === LoginStatus.REQUESTING &&
              <ActivityIndicator animating={true} style={[styles.centering, {height: 50}]} size="large" color="#3FA9F5"/>
            }
            { this.state.status === LoginStatus.LOGGED &&
              <Text style={{color: '#0C0'}}>Sesión iniciada</Text>
            }
          </View>
        </View>
      </View>
      </ScrollView>
    );
  }

  _login() {
    // Change login status
    this.setState({text: '', status: LoginStatus.REQUESTING, error: false});

    // Collect user input
    var name            = this.state.name;
    var lastName        = this.state.lastName;
    var email           = this.state.email;
    var password        = this.state.password;
    var confirmPassword = this.state.confirmPassword;

    var birthdate = this.state.birthdate;
    var gender    = this.state.gender;

    var validation = this._validate(name, lastName, email, password, confirmPassword, birthdate);
    if (!validation.result) {
      this._setError(validation.message);
      return;
    }

    var newUser = {
      email: email, 
      first_name: name, 
      last_name: lastName, 
      birthdate: birthdate, 
      gender: gender, 
      password: password
    };
    console.log(newUser);

    // Make a register request
    var register = this;

    var promise = fetch(settings.REGISTER_URL, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(newUser)
    })
    .then((response) => response.json());

    promise.then(function(result) { // ok
      console.log(result);

      if (result.code != "OK") {
        switch (result.code) {
          case 'VALIDATION_ERROR':
            register._validationError(result.error);
            break;
        }
        return;
      }

      var user = result.user;

      register.setState({status: LoginStatus.LOGGED});
      register.setState({user: user});

      AsyncStorage.setItem("user", JSON.stringify(user));

      register._goToMain();
    }, function(err) { // error
      console.log(err);
      register._setError('Error interno, inténtalo de nuevo');
    });
  }

  _validate(name, lastName, email, password, confirmPassword, birthdate) {
    if (name.length == 0 || lastName.length == 0 || email.length == 0 || password.length == 0 ||
        confirmPassword.length == 0 || birthdate === undefined)
      return {result: false, message: 'Debes llenar todos los campos'};

    if (password !== confirmPassword)
      return {result: false, message: 'Las contraseñas deben coincidir'};

    return {result: true};
  }

  _validationError(error) {
    // TODO: handle multiple exceptions at the same time

    switch (error[0].param) {
      case 'email':
        this._setError('Correo inválido');
        break;
      case 'password':
        this._setError('Contraseña inválida, al menos 4 caracteres');
        break;
      case 'first_name': 
      case 'last_name': 
      case 'birthdate': 
      case 'gender': 
        this._setError('Error interno, inténtalo de nuevo');
        break;
    }
    
  }

  _setError(message) {
    this.setState({status: LoginStatus.SLEEPING, error: message});
  }

  _backToMainMenu() {
    this.props.navigator.pop();
    return true; // This is important to prevent multiple calls
  }

  _goToMain() {
    this.props.navigator.replace({id: 'QRReader', login: {user: this.state.user}});
  }
}
