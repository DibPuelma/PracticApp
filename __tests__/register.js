import 'react-native';
import React from 'react';
import Register from '../app/routes/Register/Register'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <Register />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
