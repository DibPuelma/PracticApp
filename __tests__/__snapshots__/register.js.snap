exports[`test renders correctly 1`] = `
<ScrollView
  style={Object {}}>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 20,
        "paddingTop": 20,
      }
    }>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "maxWidth": 250,
        }
      }>
      <View
        style={
          Object {
            "backgroundColor": "#E6F4F6",
            "borderRadius": 50,
            "height": 100,
            "marginBottom": 20,
            "width": 100,
          }
        } />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#888888",
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 30,
          }
        }>
        Registro
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }>
        <TextInput
          maxLength={30}
          onChangeText={[Function]}
          placeholder="Nombre"
          style={
            Object {
              "flex": 1,
              "height": 40,
            }
          } />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }>
        <TextInput
          maxLength={30}
          onChangeText={[Function]}
          placeholder="Apellido"
          style={
            Object {
              "flex": 1,
              "height": 40,
            }
          } />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }>
        <TextInput
          maxLength={30}
          onChangeText={[Function]}
          placeholder="Correo"
          style={
            Object {
              "flex": 1,
              "height": 40,
            }
          } />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginLeft": 6,
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#999",
              "flex": 1,
              "fontSize": 12,
            }
          }>
          GÃ©nero:
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 30,
          }
        }>
        <View
          style={
            Object {
              "flex": 1,
            }
          }>
          <RCTPicker
            items={
              Array [
                Object {
                  "label": "Masculino",
                  "value": "m",
                },
                Object {
                  "label": "Femenino",
                  "value": "f",
                },
                Object {
                  "label": "Otro",
                  "value": "o",
                },
              ]
            }
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            selectedIndex={0}
            style={
              Array [
                Object {
                  "height": 216,
                },
                undefined,
              ]
            } />
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              Array [
                Object {
                  "width": 142,
                },
                Object {
                  "flex": 1,
                },
              ],
            ]
          }
          testID={undefined}>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 40,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderColor": "#aaa",
                    "borderWidth": 1,
                    "flex": 1,
                    "height": 40,
                    "justifyContent": "center",
                  },
                  Object {
                    "marginLeft": 36,
                  },
                  false,
                  false,
                ]
              }>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "#c9c9c9",
                    },
                    undefined,
                  ]
                }>
                Fecha de nacimiento
              </Text>
            </View>
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": 32,
                    "marginLeft": 5,
                    "marginRight": 5,
                    "width": 32,
                  },
                  Object {
                    "left": 0,
                    "marginLeft": 0,
                    "position": "absolute",
                    "top": 4,
                  },
                ]
              } />
            <Modal
              onRequestClose={[Function]}
              transparent={true}
              visible={false}>
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }>
                <View
                  accessibilityComponentType={undefined}
                  accessibilityLabel={undefined}
                  accessibilityTraits={undefined}
                  accessible={true}
                  hitSlop={undefined}
                  onLayout={undefined}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "alignItems": "flex-end",
                        "backgroundColor": "#00000077",
                        "flex": 1,
                        "flexDirection": "row",
                      },
                    ]
                  }
                  testID={undefined}>
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    hitSlop={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "flex": 1,
                        },
                      ]
                    }
                    testID={undefined}>
                    <View
                      style={
                        Object {
                          "backgroundColor": "#fff",
                          "height": 0,
                          "overflow": "hidden",
                        }
                      }>
                      <View
                        style={
                          Array [
                            Object {
                              "borderTopColor": "#ccc",
                              "borderTopWidth": 1,
                              "marginTop": 42,
                            },
                            undefined,
                          ]
                        }>
                        <RCTDatePicker
                          date={1262314800000}
                          maximumDate={1262314800000}
                          minimumDate={-2208978000000}
                          minuteInterval={undefined}
                          mode="date"
                          onChange={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Object {
                              "height": 216,
                            }
                          }
                          timeZoneOffsetInMinutes={undefined} />
                      </View>
                      <View
                        accessibilityComponentType={undefined}
                        accessibilityLabel={undefined}
                        accessibilityTraits={undefined}
                        accessible={true}
                        hitSlop={undefined}
                        onLayout={undefined}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "transparent",
                            },
                            Array [
                              Object {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "height": 42,
                                "justifyContent": "center",
                                "padding": 20,
                                "position": "absolute",
                                "top": 0,
                              },
                              Object {
                                "left": 0,
                              },
                              undefined,
                            ],
                          ]
                        }
                        testID={undefined}>
                        <Text
                          accessible={true}
                          allowFontScaling={true}
                          ellipsizeMode="tail"
                          style={
                            Array [
                              Object {
                                "color": "#46cf98",
                                "fontSize": 16,
                              },
                              Object {
                                "color": "#666",
                              },
                              undefined,
                            ]
                          }>
                          Cancelar
                        </Text>
                      </View>
                      <View
                        accessibilityComponentType={undefined}
                        accessibilityLabel={undefined}
                        accessibilityTraits={undefined}
                        accessible={true}
                        hitSlop={undefined}
                        onLayout={undefined}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "transparent",
                            },
                            Array [
                              Object {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "height": 42,
                                "justifyContent": "center",
                                "padding": 20,
                                "position": "absolute",
                                "top": 0,
                              },
                              Object {
                                "right": 0,
                              },
                              undefined,
                            ],
                          ]
                        }
                        testID={undefined}>
                        <Text
                          accessible={true}
                          allowFontScaling={true}
                          ellipsizeMode="tail"
                          style={
                            Array [
                              Object {
                                "color": "#46cf98",
                                "fontSize": 16,
                              },
                              undefined,
                            ]
                          }>
                          Confirmar
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </Modal>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }>
        <TextInput
          maxLength={30}
          onChangeText={[Function]}
          placeholder="ContraseÃ±a"
          secureTextEntry={true}
          style={
            Object {
              "flex": 1,
              "height": 40,
            }
          } />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }>
        <TextInput
          maxLength={30}
          onChangeText={[Function]}
          placeholder="Confirmar ContraseÃ±a"
          secureTextEntry={true}
          style={
            Object {
              "flex": 1,
              "height": 40,
            }
          } />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 10,
          }
        }>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "alignItems": "center",
                "backgroundColor": "#3FA9F5",
                "borderRadius": 5,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              },
            ]
          }
          testID={undefined}>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
                "fontSize": 15,
                "paddingBottom": 10,
                "paddingTop": 10,
              }
            }>
            ENTRAR
          </Text>
        </View>
      </View>
    </View>
  </View>
</ScrollView>
`;
